server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java

```diff
@@ -299,6 +302,15 @@ public String fieldName() {
          */
         DocValuesLoader docValuesLoader(LeafReader leafReader, int[] docIdsInLeaf) throws IOException;
 
+        /**
+         Perform any preprocessing needed before producing synthetic source.
```

Comment:
Maybe add: " to deduce whether this mapper (and its children, if any) have values to write"